File: c-programming/Examples/FromR-dotC-2/geyerex-1.c
// http://users.stat.umn.edu/~geyer/rc/
// uses GetRNGstate() to initialize the seed object.

//Note the compile command must include /usr/share/R/include
#include <R.h>
#include <Rmath.h>

void baz(int *nin, double *sin, double *tin, double *x)
{
  int n = nin[0];
  double s = sin[0];
  double t = tin[0];
  
  int i;
  
  GetRNGstate();
  
  for (i = 0; i < n; i++)
    x[i] = rbeta(s, t);
  
  PutRNGstate();
}

/*
Local Variables:
compile-command: "R CMD SHLIB geyerex-1.c -I/usr/share/R/include" 
End:
*/

File: c-programming/Examples/FromR-dotC-2/geyerex-1.R
## Make a shared object first
## R CMD SHLIB geyerex-1.c


set.seed(42)
rbeta(5, 1.5, 2.5)
dyn.load("geyerex-1.so")

baz <- function(n, s, t) {
    .C("baz", n = as.integer(n), s = as.double(s), t = as.double(t),
        x = double(n))$x
}


set.seed(42)
baz(5, 1.5, 2.5)

set.seed(42)
baz(10, 1.5, 2.5)
